core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java

```diff
@@ -275,6 +276,28 @@ protected void perform(RelOptRuleCall call, Filter filter,
     call.transformTo(relBuilder.build());
   }
 
+  /**
+   * Get conjunctions of filter's condition but with collapsed
+   * {@code IS NOT DISTINCT FROM} expressions if needed.
+   *
+   * @param filter filter containing condition
+   * @return condition conjunctions with collapsed {@code IS NOT DISTINCT FROM}
+   * expressions if any
+   * @see RelOptUtil#conjunctions(RexNode)
+   */
+  private List<RexNode> getConjunctions(Filter filter) {
```

Comment:
For the Join condition, it seems like we should always remember for `collapseExpandedIsNotDistinctFromExpr`. If new rule added, users should do this expansion again, just like this PR and https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java#L92
I mean that shall we extract such behavior as a shared `RexShuttle`, and have an approach to make this expansion automatic ?

Reply:
This returns conjunctions of filer condition, and if only 2 places use this code, there is no necessity to have a common `RexShuttle`.